### ISSUE TITLE
Complex JSONPath breaking between v12.0.3 and v13.0.1

### ISSUE THREAD

**Issue by lauriekrueger** [JamesNK/Newtonsoft.Json#2987]
My complex JSONPath queries that worked with version 12 and below no longer work with version 13+

### Source/destination JSON
{
    "id": 21,
    "name": "Accrual Profiles",
    "uniqueKey": "WSAAccrualProfile",
    "itemsRetrieveResponseDTOs": [],
    "itemsRetrieveResponses": [
        {
            "itemDataInfo": {
                "title": "-TIL",
                "key": "-TIL",
                "env": null,
                "urlparams": "Name=-TIL"
            },
            "responseObjectNode": {
                "@Status": "Success",
                "@Action": "RetrieveForUpdate",
                "WSAAccrualProfile": {
                    "@Name": "-TIL",
                    "AccrualProfilePolicies": {
                        "WSAAccrualProfilePolicy": [
                            {
                                "@AccrualPolicy": "-Time In Lieu",
                                "@Display": true
                            },
                            {
                                "@AccrualPolicy": "-Time In Lieu 2",
                                "@Display": true
                            }
                        ]
                    },
                    "@PersistentId": "-TIL",
                    "AccrualProfileBalanceCascadeGroups": {
                        "WSAAccrualProfileBalanceCascadeGroup": {
                            "@BalanceCascadeGroup": "-TIL"
                        }
                    }
                }
            }
        }
    ]
}

### Expected behavior

I expect to get a IEnumberable<JToken> from calling SelectTokens(path) with this path

 $..['WSAAccrualProfile'][?(@['@Name']=='-TIL')]

the above path works in v12.0.3

### Actual behavior

null is returned from SelectTokens(path)

### Steps to reproduce

        public IEnumerable<JToken> SelectTokens(string xpath)
        {
            return _token.SelectTokens(xpath);
        }

**Comment by elgonzo** [comment]
> null is returned from SelectTokens(path)

It should return an empty enumerable, because the correct result of applying your JsonPath expression to your given json data example should be an empty result.

If v12.0.3 produces anything other than an empty result for your given json example and JsonPath expression, then the bug is in v12.0.3. Applying your given JsonPath expression to your json example must give an empty result according to the JsonPath specification. (Your JsonPath expression looks for _children_ of `WSAAccrualProfile` which have some `@name` value. But no child of `WSAAccrualProfile` has a `@name` value, only `WSAAccrualProfile` itself has. Hence why the correct result for the given JsonPath expression must be empty here.)

So, basically v13.0.x yielding an empty result is the correct result.

You need to fix your JsonPath expression. On possible JsonPath expression that would select what you want is:
```
$..[?(@.WSAAccrualProfile['@name'] == '-TIL')].WSAAccrualProfile
```
It selects any object that has a `WSAAccrualProfile` child object with a `@name` value of "-TIL" and then selects the `WSAAccrualProfile` child from that object. (Note that i haven't tested this with Newtonsoft.Json. I am not using Newtonsoft.Json anymore and its limited JsonPath support is based on an extremely old JsonPath spec, and i haven't verified if the JsonPath i suggested isn't using any features not yet available in that old JsonPath spec, so YMMV...)
